src/components/Audio.cpp

```diff
@@ -178,51 +178,115 @@ bool Audio::setRate(double rate)
-  int16_t* output = (int16_t*)alloca(out_len * 2);
-
-  if (output == NULL)
+  if (_sampleRate == _coreRate)
   {
-    _logger->error(TAG "Error allocating output buffer");
-    return;
+    /* no resampling needed */
+    output = (int16_t*)samples;
+    out_len = frames * 2;
+    output_size = frames * 2 * sizeof(int16_t);
   }
-
-  _logger->debug(TAG "Resampling %u samples to %u", in_len, out_len);
-  int error = speex_resampler_process_int(_resampler, 0, samples, &in_len, output, &out_len);
-  _logger->debug(TAG "Resampled  %u samples to %u", in_len, out_len);
-
-  if (error != RESAMPLER_ERR_SUCCESS)
+  else
   {
-    memset(output, 0, out_len * 2);
-    _logger->error(TAG "speex_resampler_process_int: %s", speex_resampler_strerror(error));
+    // /* adjust the audio input rate. */
+    //const int    half_size = (int)_fifo->size() / 2;
+    //const int    delta_mid = (int)avail - half_size;
+    //const double direction = (double)delta_mid / (double)half_size;
+    //const double rateControlDelta = 0.005;
+    //const double adjust = 1.0 + rateControlDelta * direction;
+
+    //_currentRatio = _originalRatio * adjust;
+
+    //if (_currentRatio != _originalRatio)
+    //  _logger->debug(TAG "Original ratio %f adjusted by %f to %f", _originalRatio, adjust, _currentRatio);
+
+    /* allocate output buffer */
+    out_len = (spx_uint32_t)(frames * 2 * _currentRatio);
+    out_len += (out_len & 1);  /* don't send incomplete frames */
+
+    output_size = out_len * sizeof(int16_t);
+    output = (int16_t*)alloca(output_size);
+
+    if (output == NULL)
+    {
+      _logger->error(TAG "Error allocating audio resampling output buffer");
+      return;
+    }
+
+    /* do the resampling */
+    spx_uint32_t in_len = frames * 2; /* input is 2-channel */
+    _logger->debug(TAG "Resampling %u samples to %u", in_len, out_len);
+    int error = speex_resampler_process_int(_resampler, 0, samples, &in_len, output, &out_len);
+
+    if (error != RESAMPLER_ERR_SUCCESS)
+    {
+      memset(output, 0, output_size);
+      _logger->error(TAG "speex_resampler_process_int: %s", speex_resampler_strerror(error));
+    }
   }
 
-  out_len &= ~1; // don't send incomplete audio frames
-  size_t size = out_len * 2;
-  
-  while (size > avail)
+  /* flush to FIFO queue */
+  const size_t needed = (output_size / 2) * _channels;
+  while (avail < needed)
   {
-    _logger->debug(TAG "Requested %zu bytes but only %zu available, sleeping", size, avail);
+    _logger->debug(TAG "Waiting for FIFO (need %zu bytes but only %zu available), sleeping", needed, avail);
     SDL_Delay(1);
     avail = _fifo->free();
   }
 
-  _fifo->write(output, size);
-  _logger->debug(TAG "Wrote %zu bytes to the FIFO", size);
+  if (_channels == 2)
+  {
+    /* input is 2 channels, output is 2 channels, just flush it */
+    _fifo->write(output, output_size);
+  }
+  else
+  {
+    /* input is 2 channels, have to add silence for other channels */
```

Comment:
I always thought that SDL would take care of everything when the requested audio spec could not be matched by the hardware, including channel conversion. Maybe it's initialized or used differently here in RALibretro, I don't remember.